File: Visitor/python-source/rooms/ElementRoom.py
from __future__ import annotations

from abc import ABC
from abc import abstractmethod
from typing import TYPE_CHECKING

from .Surface import Surface

if TYPE_CHECKING:
    from workers import Visitor


class ElementRoom(ABC):
    """Комната, к которой может прийти рабочий.

    Метод accept здесь единственный обязательный, и он всегда написан одинаково:
    комната зовёт у гостя тот метод, который соответствует её типу. Такой приём
    называют двойной диспетчеризацией: выбор кода зависит и от типа комнаты, и
    от типа рабочего.
    """

    def __init__(self) -> None:
        self.wall = Surface()
        self.floor = Surface()
        self.ceiling = Surface()

    @abstractmethod
    def accept(self, visitor: Visitor) -> None:
        raise NotImplementedError

    @abstractmethod
    def get_name(self) -> str:
        raise NotImplementedError

    def show_decorations(self) -> None:
        print(f"{self.get_name()}:")
        print(f"  wall:    {self.wall}")
        print(f"  floor:   {self.floor}")
        print(f"  ceiling: {self.ceiling}")

File: JavaSource/src/main/java/ru/akhcheck/patterns/visitor/rooms/ElementRoom.java
package ru.akhcheck.patterns.visitor.rooms;

import ru.akhcheck.patterns.visitor.workers.Visitor;

/**
 * Комната, к которой может прийти рабочий.
 *
 * <p>Метод accept всегда написан одинаково: комната зовёт у гостя тот метод,
 * который соответствует её типу. Такой приём называют двойной диспетчеризацией,
 * потому что выбор кода зависит и от типа комнаты, и от типа рабочего.
 */
public abstract class ElementRoom {
    private final Surface wall = new Surface();
    private final Surface floor = new Surface();
    private final Surface ceiling = new Surface();

    public abstract void accept(Visitor visitor);

    public abstract String getName();

    public Surface getWall() {
        return wall;
    }

    public Surface getFloor() {
        return floor;
    }

    public Surface getCeiling() {
        return ceiling;
    }

    public void showDecorations() {
        System.out.println(getName() + ":");
        System.out.println("  wall:    " + wall);
        System.out.println("  floor:   " + floor);
        System.out.println("  ceiling: " + ceiling);
    }
}
